add SimpleQueryTest CQL testsuite

ananyaks opened the pull request:
This pull request is for adding SimpleQueryTest suite to Yugabyte-DB's CQL testsuite. I have successfully run the first 7 tests of Cassandra's CQL test suite SimpleQueryTest.java (under Apache license) with Yugabyte-DB. The changes I had to make are as follows:

**SimpleQueryTest.java**  

This is a testsuite with many individual tests. The first 7 tests currently run fine, the rest are commented out for now. Goal was to keep the changes to this file to the minimum so that new individual tests can be run easily. The changes to this file are:

- replaced the inclusion of cassandra package with org.yb.cql
- calls to flush() has been commented out, as I do not know the equivalent in Yugabyte-DB. The flush() in Cassandra's testsuite is a public method of CQLTester base class that calls the ColumnFamilyStore method forceBlockingFlush()

**CQLTester.java**
The class SimpleQueryTest inherits from CQLTester. I modified CQLTester.java (also under Apache license) much more to only include the minimum necessary to successfully compile and run SimpleQueryTest. These modifications were important for  keeping the changes in SimpleQueryTest to be minimum. The CQLTester.java changes are:

- import yugabyte-db relevant classes/package
- made CQLTester a sub-class of BaseCQLTest to pull in yugabyte-db boilerplate
- modified createTable and execute methods to provide the behavior expected by the individual tests in SimpleQueryTest. For example, the individual tests do not hardcode the table name instead it gets generated. During execution of statements (e.g. select, insert) the CQL statement string gets built with the looked up table name
- modified assertRows() methods which is an easy way to verify the rows inserted/deleted match the rows selected. This method required more modifications
- a comparison with Cassandra's CQLTester.java can be used to understand the modifications

Thanks!

kmuthukk commented:
I think explicit calls to flush(), disableCompactions() etc. can be removed (as that's more for white-box testing).

Thanks for submitting this pull request @ananyaks . We will try to review this and merge to code base.

ananyaks commented:
Hi @kmuthukk, @robertpang, I did another commit by un-commenting `testRangeTombstones()` and looks like this test is not working because of a bug in DELETE statement execution. There are 300 rows inserted into the table before executing the DELETE statement. The DELETE statment was supposed to only delete 100 rows that matched `k = "key" AND t1 = 1`, instead it deleted all 300 rows. It looks like the `AND `in the DELETE statement was treated as an `OR `and since all rows match `k="key"` the DELETE statement deleted all rows. This looks like a bug.
Thanks!

kmuthukk commented:
Thanks for reporting @ananyaks!

I have filed https://github.com/YugaByte/yugabyte-db/issues/434 for the issue you found.

robertpang commented:
@ananyaks Thank you for your contribution to YugaByte! I will start the review right away.

kmuthukk commented:
Hi @ananyaks : Issue #434 has been fixed (https://github.com/YugaByte/yugabyte-db/commit/b2164ff7f98bf5449083701382e3daa12bd55c05).

ananyaks commented:
Thanks @kmuthukk  for letting me know. I am not able to test the fixes because:

I have been using the local repository that was cloned directly from [yugabyte-db](https://github.com/YugaByte/yugabyte-db) version in July and successfully built also in July. In this local repository I am able to run SimpleQueryTest.java test suite.

To submit the pull request, I recently did a [fork of yugabyte-db](https://github.com/ananyaks/yugabyte-db) in my github and cloned this fork into a different local repository. I created a branch and then committed the SimpleQueryTest.java and CQLTester.java into this branch and did a pull request from this branch using my github. In order to test with the latest version of [yugabyte-db](https://github.com/YugaByte/yugabyte-db), I tried to build but ran into build errors. The file [buildErrors.txt](https://github.com/YugaByte/yugabyte-db/files/2285622/buildErrors.txt)
shows the errors I am getting. Should I file this as a new configure error issue?
Thanks!

mbautin commented:
@ananyaks Can you try rebasing and building again? We recently fixed this build issue with the AWS C++ SDK due to incorrect RPATH rewriting that we were doing.

mbautin commented:
@ananyaks: we appreciate your contribution and we will complete the review of your patch as soon as possible. In the meantime, could you please sign our CLA (Contributor License Agreement) at https://bit.ly/sign-yugabyte-cla? Thank you!

robertpang commented:
@ananyaks Just a quick note that I am reviewing your PR.

ananyaks commented:
Hi @mbautin, I have signed the CLA.
Thanks!

ananyaks commented:
Hi @robertpang, thanks for reviewing my PR. Let me know if I need to make any changes.
Thanks!

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/CQLTester.java:
```diff
@@ -0,0 +1,216 @@
+            return;
+        }
+	ColumnDefinitions colDef = result.getColumnDefinitions();
+	List<ColumnDefinitions.Definition> meta = colDef.asList();
```
nit: alignment

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/CQLTester.java:
```diff
@@ -0,0 +1,216 @@
+		  case INT:
+		    int intActual = actual.getInt(j);
+		    int intExpected = (Integer)expected[j];
+		    assertEquals(intActual, intExpected);
```
May read better like below.

`assertEquals((Integer)expected[j], actual.getInt(j));`

Also, note that in JUnit API, the parameters of the assert methods should be `assertXXX(expected, actual)`.

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/CQLTester.java:
```diff
@@ -0,0 +1,216 @@
+		    String strActual = actual.getString(j);
+		    String strExpected = (String)expected[j];
+		    assertEquals(strActual, strExpected);
+		    LOG.info("Row[" + i + ", " + j + "] = " + strActual);
```
Ditto.

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/CQLTester.java:
```diff
@@ -0,0 +1,216 @@
+        }
+
+        if (iter.hasNext())
+	  fail("Got more rows than expected");
```
Alignment.

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/CQLTester.java:
```diff
@@ -0,0 +1,216 @@
+        if (iter.hasNext())
+	  fail("Got more rows than expected");
+        if (i != rows.length)
+	  fail("Got less rows than expected");
```
Alignment.

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/SimpleQueryTest.java:
```diff
@@ -0,0 +1,572 @@
+        execute("INSERT INTO %s (k, t, v) values (?, ?, ?)", "key", 2, "v12");
+        execute("INSERT INTO %s (k, t, v) values (?, ?, ?)", "key", 3, "v13");
+
+        // flush();
```
Let's remove all `flush()` calls altogether.

ananyaks commented:
@mbautin , I was able to build without errors after rebasing.

@kmuthukk, I tested `testRangeTombstones()` and with the fix  ([b2164ff](https://github.com/YugaByte/yugabyte-db/commit/b2164ff7f98bf5449083701382e3daa12bd55c05)) it worked without any asserts.

@robertpang, I will make the changes you have asked.

Thanks!

kmuthukk commented:
Excellent. Thanks a lot @ananyaks for the updates.

ananyaks commented:
Hi @robertpang, Thanks for your suggestions, I have done the changes. Please check [961afc2](https://github.com/YugaByte/yugabyte-db/pull/429/commits/961afc23fd8b32f65fdc54c6498f3ef7ab6bd29e) and [5728ff2](https://github.com/YugaByte/yugabyte-db/pull/429/commits/5728ff2e60a4ec291005dfa22dbb43c203402555)

Thanks!

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/CQLTester.java:
```diff
@@ -0,0 +1,187 @@
+        Iterator<Row> iter = result.iterator();
+        int i = 0;
+        if (!iter.hasNext() && rows.length > 0)
+	    fail(String.format("No rows returned by query but %d expected", rows.length));
```
nit: indent by 4 spaces.

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/CQLTester.java:
```diff
@@ -0,0 +1,187 @@
+        List<ColumnDefinitions.Definition> meta = colDef.asList();
+        Iterator<Row> iter = result.iterator();
+        int i = 0;
+        if (!iter.hasNext() && rows.length > 0)
```
This check is unnecessary. The check `if (i != rows.length) ...` after the `while` loop below does the same check essentially.

ananyaks replied:
Thanks @robertpang. This check was raising "No rows returned by query..." error, to not loose this error case, I changed the failure message for the `if (i != rows.length)` to `fail(String.format("Query returned %d rows but expected %d rows", i, rows.length));`

Thanks!

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/CQLTester.java:
```diff
@@ -0,0 +1,187 @@
+                switch (type) {
+                    case INT:
+                        assertEquals((Integer)expected[j], (Integer)actual.getInt(j));
+                        LOG.info("Row[{}, {}] = {}", i, j, actual.getInt(j));
```
Please log before the `assertEquals`. Otherwise when assert fails, it aborts and we won't have this log message for the assert failure. Ditto for the same instance below.

ananyaks commented:
Hi @kmuthukk , @robertpang , after my last night's commit of [changes due to more reviewer comments](https://github.com/YugaByte/yugabyte-db/pull/429/commits/0dfeb2aa19994cda2fffda65bc51bc75739d98f8), I did another commit to [add 3 more individual tests](https://github.com/YugaByte/yugabyte-db/pull/429/commits/c0172571cb9ac87b063718e70664f4944d8eff21) of SimpleQueryTest testsuite, running these 3 found the following issues:

**Test test2ndaryIndexes():**
After creating the below table without error:
`CREATE TABLE cql_test_keyspace.table_3 (k text, t int, v text, PRIMARY KEY (k, t));`

Create index failed due to syntax error:
```
com.datastax.driver.core.exceptions.SyntaxError: SQL error: Feature Not Supported
CREATE INDEX ON cql_test_keyspace.table_3(v)

```
**Test testStaticColumns():**
`[ERROR]   SimpleQueryTest.testStaticColumns:379->CQLTester.assertRows:191 expected:<st5> but was:<null>`

**Test testDistinct():**
`[ERROR]   SimpleQueryTest.testDistinct:395->CQLTester.assertRows:191 expected:<key[1]> but was:<key[2]>`

Thanks!

robertpang commented:
Thanks, @ananyaks.

The first issue with test2ndaryIndexes() is expected. Currently we require the index to be created with a name explicitly. We should support auto-generating the index name in future but for now can you add an index name (akin to how the test harness auto-generate the table name)?

We will look into the second and third issues.

robertpang commented:
@ananyaks BTW, per our Java coding convention, we indent by 2 spaces instead of 4 at each level. May I ask you update the indentation? Thanks.

ananyaks commented:
Hi @kmuthukk, @robertpang, after my last commit of [indent 2 instead of 4 spaces and generate index name](https://github.com/YugaByte/yugabyte-db/pull/429/commits/575cf96e2b942888bb44cb295ee5aabb7a60e186) I just now committed [run all tests in SimpleQueryTest and support set column type in assertRows()](https://github.com/YugaByte/yugabyte-db/pull/429/commits/f2a307d3664005e01afa3e92fe8f009352f17881) which runs all 18 individual tests in SimpleQueryTest suite. The good news is that there are only 2 errors (the same ones in my previous post) and the other 16 tests run successfully! 
```
[ERROR] Failures: 
[ERROR]   SimpleQueryTest.testDistinct:397->CQLTester.assertRows:247 expected:<key[1]> but was:<key[2]>
[ERROR]   SimpleQueryTest.testStaticColumns:380->CQLTester.assertRows:247 expected:<st5> but was:<null>
[INFO] 
[ERROR] Tests run: 18, Failures: 2, Errors: 0, Skipped: 0
```
The `testDistinct `is failing because the order of rows is switched. The test is expecting:
```
    assertRows(execute("SELECT DISTINCT k FROM %s"),
      row("key1"),
      row("key2")
```
But yugabyte-db is returning in the below order. Is this a yugabyte-db behavior and not a bug?
```
    assertRows(execute("SELECT DISTINCT k FROM %s"),
      row("key2"),
      row("key1")
```
The `testStaticColumns `issue seems to be a bug. 
```
    assertRows(execute("SELECT * FROM %s WHERE k = ? AND t = ?", "key1", 2),
      row("key1", 2, "st5", "foo2"));
```
The above assert succeeds.
```
    displayRows(execute("SELECT * FROM %s WHERE k = ? AND t = ? ORDER BY t DESC", "key1", 2));
Displays: Row 0 = Row[key1, 2, NULL, foo2]
```
So the below assert fails since col2 is `NULL `instead of `"st5"`. Looks like` "ORDER BY t DESC"` handling in `SELECT `maybe causing the issue?
```
    assertRows(execute("SELECT * FROM %s WHERE k = ? AND t = ? ORDER BY t DESC", "key1", 2),
      row("key1", 2, "st5", "foo2"));
```
The difference with cassandra sql behavior that I have worked around are:
1. After creating the below table without error:
`CREATE TABLE cql_test_keyspace.table_3 (k text, t int, v text, PRIMARY KEY (k, t));`
Create index failed due to syntax error:
```
com.datastax.driver.core.exceptions.SyntaxError: SQL error: Feature Not Supported
CREATE INDEX ON cql_test_keyspace.table_3(v)
```
Based on @robertpang feedback, generated the index name by adding `createIndex() `methods in `CQLTester.java`. Modified the test as:
`    createIndex("CREATE INDEX %s ON %s(v);");`

2. The second difference with cassandra sql is
```
Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: SQL 
error: Invalid Table Definition. Transactions cannot be enabled in an 
index of a table without transactions enabled.
CREATE INDEX cql_test_keyspace_index_4 ON cql_test_keyspace.table_3(v);
^^^^^^
```
To resolve the above bug I found the yugabyte documentation on [create index](https://docs.yugabyte.com/latest/api/cassandra/ddl_create_index/) said the table created should have this `WITH transactions = { 'enabled' : true };`
I appended the above to the create table statement and the test worked.

Thanks!

ananyaks commented:
Hi @kmuthukk, @robertpang, continuing my previous post...

Some methods that I have implemented in CQLTester.java that you may find useful:

1. `assertRows() `whose example usage is below. It is easy to assert the output of a `SELECT`
```
    assertRows(execute("SELECT * FROM %s WHERE v = ?", "foo"),
      row("key1",  1, "foo"),
      row("key2",  1, "foo"),
      row("key2",  2, "foo")
    );
```
I also had to support columns of type `set`. `assertRows()` can be extended to support other column types.  
```
    assertRows(execute("SELECT s FROM %s WHERE k = ?", 1),
      row(set(2))
```
I fixed `int `columns that can have `null `so one can do:
```
    assertRows(execute("SELECT * FROM %s"),
      row(1, 1, 1, 1),
      row(2, 2, 2, 2),
      row(3, null, 3, null)
    );
```
2. `displayRows()` that can log a `ResultSet` as in
    `displayRows(execute("SELECT * FROM %s WHERE k = ? AND t = ? ORDER BY t DESC", "key1", 2));`
Displays: 
     `Row 0 = Row[key1, 2, NULL, foo2]`

Thanks!

robertpang commented:
@ananyaks Sorry for my delay to get back to you.

> Looks like "." is not allowed in index name (even though it is allowed in table name), so I replaced with "_" in the generated name. Is this a bug or works as expected?

This is the expected behavior. An index is always created in the same keyspace as the table it indexes. For that reason, the index name is required to be a simple name without "."  and keyspace name.

> To resolve the above bug I found the yugabyte documentation on create index said the table created should have this
WITH transactions = { 'enabled' : true };
I appended the above to the create table and the test worked. This is a difference with Cassandra CQL as there the table created did not need WITH transacations ...

This is not a bug. Unlike Apache Cassandra, indexes in YugaByte DB are strongly consistent and for that reason the table being indexed should have transactions enabled.

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/CQLTester.java:
```diff
@@ -28,180 +28,267 @@
-    }
+  protected void createIndex(String keyspace, String query)
+  {
+    String fullQuery = String.format(query, keyspace + "_" + createIndexName(), keyspace + "." + currentTable());
```
As explained separately, index name should be a simple name without keyspace. Please remove `keyspace` and `_` from the index name.

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/SimpleQueryTest.java:
```diff
@@ -0,0 +1,540 @@
+    );
+  }
+
+  @Test
```
Please file a Github issue on this test failure and add `@Ignore ("Disabled for issue #<issue number>")` to temporarily disable this test. We will continue to look into the issue.

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/SimpleQueryTest.java:
```diff
@@ -0,0 +1,540 @@
+      row("key1", 2, "st5", "foo2"));
+  }
+
+  @Test
```
Please file a Github issue on this test failure and add `@Ignore ("Disabled for issue #<issue number>")` to temporarily disable this test. We will continue to look into the issue.

ananyaks commented:
Hi @robertpang, Thank you for your feedback! I have incorporated them in the commit [updated as per comments and ignored 2 test cases that run into issues 465 and 466](https://github.com/YugaByte/yugabyte-db/pull/429/commits/e8286beffd6aac4efa994a588b57fff541028063). I have created issues [465 ](https://github.com/YugaByte/yugabyte-db/issues/465)and [466 ](https://github.com/YugaByte/yugabyte-db/issues/466
)and inserted `@Ignore` before the test cases `testStaticColumns()` and `testDistinct()`. All remaining 16 tests in SimpleQueryTest suite run successfully and I get the following output:

```
[INFO] Running org.yb.cql.SimpleQueryTest
[WARNING] Tests run: 18, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 155.894 s - in org.yb.cql.SimpleQueryTest
[INFO] Results:
[INFO] 
[WARNING] Tests run: 18, Failures: 0, Errors: 0, Skipped: 2
```
Thanks!

robertpang commented on java/yb-cql/src/test/java/org/yb/cql/SimpleQueryTest.java:
```diff
@@ -0,0 +1,532 @@
+  }
+
+  @Test
+  @Ignore ("Disabled for issue #466")
```
@ananyaks I had a closer look of this issue and this is actually not a bug. Please see #466 for my explanation. Can you reverse the order of the expected rows and re-enable this test case? Sorry for that.

ananyaks commented:
Hi @robertpang, 

Yes I also though so and had questioned whether it was a bug. I have removed the `@Ignore` for `testDistinct()` in the latest [commit](https://github.com/YugaByte/yugabyte-db/pull/429/commits/dc98351147bc8af020fc73f402f59bd493117f45). 

Instead of reversing the order of the expected rows, I first tried to get sorted order of the result by using an ORDER BY as below. I thought that this would make the results not depend on the hash algorithm.

`SELECT DISTINCT k FROM cql_test_keyspace.table_15 ORDER BY k ASC;
`
But the test failed with:
```
[ERROR] testDistinct(org.yb.cql.SimpleQueryTest)  Time elapsed: 6.153 s  <<< ERROR!
com.datastax.driver.core.exceptions.InvalidQueryException: 
SQL error: Invalid Arguments. All hash columns must be set if order by clause is present.
SELECT DISTINCT k FROM cql_test_keyspace.table_15 ORDER BY k ASC
                                                           ^^^^^
 (error -304)
```
How should "_all hash columns be set_" by the query? Or, is this a bug that I should file as an issue?

I have gone ahead and reversed the order of expected rows in the recent commit to pass testDistinct(). Now, other than one ignored test (`testStaticColumns()`), all other 17 test pass!

Thanks!

robertpang commented:
Thanks to @ananyaks for your contribution to YugaByte. This PR has been merged in commit [2653e9b](https://github.com/YugaByte/yugabyte-db/commit/2653e9bd7fb70f36718c6db742a7f40edb86bcf0). The disabled test has also been enabled given the issue #465 has been fixed.
